In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split

from sklearn.ensemble import RandomForestRegressor

from sklearn.preprocessing import RobustScaler 

from sklearn.metrics import mean_squared_error

# 🏆 9o diamonds (sin GridSearchCV)
##  RSME = 544
## Score competi = 547

In [2]:
diamonds_1 = pd.read_csv('./data/diamonds_1.csv')
diamonds_2 = pd.read_csv('./data/diamonds_2.csv')
diamonds_3 = pd.read_csv('./data/diamonds_3.csv')
diamonds_4 = pd.read_csv('./data/diamonds_4.csv')
diamonds_5 = pd.read_csv('./data/diamonds_5.csv')
diamonds_6 = pd.read_csv('./data/diamonds_6.csv')
diamonds_7 = pd.read_csv('./data/diamonds_7.csv')
diamonds_8 = pd.read_csv('./data/diamonds_8.csv')
diamonds_9 = pd.read_csv('./data/diamonds_9.csv')

In [4]:
print(diamonds_1.shape)
print(diamonds_2.shape)
print(diamonds_3.shape)
print(diamonds_4.shape)
print(diamonds_5.shape)
print(diamonds_6.shape)
print(diamonds_7.shape)
print(diamonds_8.shape)
print(diamonds_9.shape)

(40439, 26)
(40439, 9)
(40455, 9)
(40439, 7)
(40439, 21)
(40439, 19)
(40439, 10)
(40439, 36)
(40455, 10)


In [3]:
X_1 = diamonds_1.drop('price', axis=1)
y_1 = diamonds_1['price']

X_2 = diamonds_2.drop('price', axis=1)
y_2 = diamonds_2['price']

X_3 = diamonds_3.drop('price', axis=1)
y_3 = diamonds_3['price']

X_4 = diamonds_4.drop('price', axis=1)
y_4 = diamonds_4['price']

X_5 = diamonds_5.drop('price', axis=1)
y_5 = diamonds_5['price']

X_6 = diamonds_6.drop('price', axis=1)
y_6 = diamonds_6['price']

X_7 = diamonds_7.drop('price', axis=1)
y_7 = diamonds_7['price']

X_8 = diamonds_8.drop('price', axis=1)
y_8 = diamonds_8['price']

X_9 = diamonds_9.drop('price', axis=1)
y_9 = diamonds_9['price']

# ⚠️ NO SCALER ⚠️

In [4]:
X_1_train, X_1_test, y_1_train, y_1_test = train_test_split(X_1, y_1, test_size=0.2, random_state=42)

X_2_train, X_2_test, y_2_train, y_2_test = train_test_split(X_2, y_2, test_size=0.2, random_state=42)
X_3_train, X_3_test, y_3_train, y_3_test = train_test_split(X_3, y_3, test_size=0.2, random_state=42)
X_4_train, X_4_test, y_4_train, y_4_test = train_test_split(X_4, y_4, test_size=0.2, random_state=42)
X_5_train, X_5_test, y_5_train, y_5_test = train_test_split(X_5, y_5, test_size=0.2, random_state=42)
X_6_train, X_6_test, y_6_train, y_6_test = train_test_split(X_6, y_6, test_size=0.2, random_state=42)
X_7_train, X_7_test, y_7_train, y_7_test = train_test_split(X_7, y_7, test_size=0.2, random_state=42)
X_8_train, X_8_test, y_8_train, y_8_test = train_test_split(X_8, y_8, test_size=0.2, random_state=42)
X_9_train, X_9_test, y_9_train, y_9_test = train_test_split(X_9, y_9, test_size=0.2, random_state=42)

In [44]:
model = RandomForestRegressor()

In [45]:
# model.fit(X_1_train, y_1_train)
# model.fit(X_2_train, y_2_train)
# model.fit(X_3_train, y_3_train)
# model.fit(X_4_train, y_4_train)
# model.fit(X_5_train, y_5_train)
# model.fit(X_6_train, y_6_train)
# model.fit(X_7_train, y_7_train)
# model.fit(X_8_train, y_8_train)
model.fit(X_9_train, y_9_train)

RandomForestRegressor()

In [46]:
# pred_1 = model.predict(X_1_test)
# pred_2 = model.predict(X_2_test)
# pred_3 = model.predict(X_3_test)
# pred_4 = model.predict(X_4_test)
# pred_5 = model.predict(X_5_test)
# pred_6 = model.predict(X_6_test)
# pred_7 = model.predict(X_7_test)
# pred_8 = model.predict(X_8_test)
pred_9 = model.predict(X_9_test)


In [47]:
# rmse_1 = mean_squared_error(y_1_test, pred_1, squared=False)
# rmse_2 = mean_squared_error(y_2_test, pred_2, squared=False)
# rmse_3 = mean_squared_error(y_3_test, pred_3, squared=False)
# rmse_4 = mean_squared_error(y_4_test, pred_4, squared=False)
# rmse_5 = mean_squared_error(y_5_test, pred_5, squared=False)
# rmse_6 = mean_squared_error(y_6_test, pred_6, squared=False)
# rmse_7 = mean_squared_error(y_7_test, pred_7, squared=False)
# rmse_8 = mean_squared_error(y_8_test, pred_8, squared=False)
rmse_9 = mean_squared_error(y_9_test, pred_9, squared=False)

C:\Users\tete1\miniconda3\envs\jupyter_env\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


In [48]:
print(rmse_1)
print(rmse_2)
print(rmse_3)
print(rmse_4)
print(rmse_5)
print(rmse_6)
print(rmse_7)
print(rmse_8)
print(rmse_9)

634.1119701899296
549.1983003898127
545.160971230846
563.6113096877659
621.5643605385723
560.295547552754
551.5032177407044
643.9101382274771
544.2653108311806


## 🏅 Winner: 9o diamonds RSME = 544.26

# ⚠️ ROBUST SCALER ⚠️

In [78]:
robust_scaler = RobustScaler()

In [79]:
# X_1_rob = robust_scaler.fit_transform(X_1)
# X_2_rob = robust_scaler.fit_transform(X_2)
# X_3_rob = robust_scaler.fit_transform(X_3)
# X_4_rob = robust_scaler.fit_transform(X_4)
# X_5_rob = robust_scaler.fit_transform(X_5)
# X_6_rob = robust_scaler.fit_transform(X_6)
X_7_rob = robust_scaler.fit_transform(X_7)

In [80]:
# X_1_rob_train, X_1_rob_test, y_1_rob_train, y_1_rob_test = train_test_split(X_1_rob, y_1, test_size=0.2, random_state=42)

# X_2_rob_train, X_2_rob_test, y_2_rob_train, y_2_rob_test = train_test_split(X_2_rob, y_2, test_size=0.2, random_state=42)
# X_3_rob_train, X_3_rob_test, y_3_rob_train, y_3_rob_test = train_test_split(X_3_rob, y_3, test_size=0.2, random_state=42)
# X_4_rob_train, X_4_rob_test, y_4_rob_train, y_4_rob_test = train_test_split(X_4_rob, y_4, test_size=0.2, random_state=42)
# X_5_rob_train, X_5_rob_test, y_5_rob_train, y_5_rob_test = train_test_split(X_5_rob, y_5, test_size=0.2, random_state=42)
# X_6_rob_train, X_6_rob_test, y_6_rob_train, y_6_rob_test = train_test_split(X_6_rob, y_6, test_size=0.2, random_state=42)
X_7_rob_train, X_7_rob_test, y_7_rob_train, y_7_rob_test = train_test_split(X_7_rob, y_7, test_size=0.2, random_state=42)

In [81]:
model = RandomForestRegressor()

In [82]:
# model.fit(X_1_rob_train, y_1_rob_train)
# model.fit(X_2_rob_train, y_2_rob_train)
# model.fit(X_3_rob_train, y_3_rob_train)
# model.fit(X_4_rob_train, y_4_rob_train)
# model.fit(X_5_rob_train, y_5_rob_train)
# model.fit(X_6_rob_train, y_6_rob_train)
model.fit(X_7_rob_train, y_7_rob_train)

RandomForestRegressor()

In [83]:
# pred_1_rob = model.predict(X_1_rob_test)
# pred_2_rob = model.predict(X_2_rob_test)
# pred_3_rob = model.predict(X_3_rob_test)
# pred_4_rob = model.predict(X_4_rob_test)
# pred_5_rob = model.predict(X_5_rob_test)
# pred_6_rob = model.predict(X_6_rob_test)
pred_7_rob = model.predict(X_7_rob_test)

In [84]:
# rmse_1_rob = mean_squared_error(y_1_rob_test, pred_1_rob, squared=False)
# rmse_2_rob = mean_squared_error(y_2_rob_test, pred_2_rob, squared=False)
# rmse_3_rob = mean_squared_error(y_3_rob_test, pred_3_rob, squared=False)
# rmse_4_rob = mean_squared_error(y_4_rob_test, pred_4_rob, squared=False)
# rmse_5_rob = mean_squared_error(y_5_rob_test, pred_5_rob, squared=False)
# rmse_6_rob = mean_squared_error(y_6_rob_test, pred_6_rob, squared=False)
rmse_7_rob = mean_squared_error(y_7_rob_test, pred_7_rob, squared=False)

C:\Users\tete1\miniconda3\envs\jupyter_env\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


In [85]:
print(rmse_1_rob)
print(rmse_2_rob)
print(rmse_3_rob)
print(rmse_4_rob)
print(rmse_5_rob)
print(rmse_6_rob)
print(rmse_7_rob)

634.4082896385572
549.0708993453093
637.3408548852266
562.2912438468542
624.5288474670245
562.641901251339
553.2234249191879


## 🏅 Winner: 2do diamonds RSME = 549.07

# Now: Diamonds_test + NO Scaler

In [16]:
diamonds_test = pd.read_csv('./data/diamonds_test.csv')

In [50]:
diamonds_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13485 entries, 0 to 13484
Data columns (total 11 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       13485 non-null  int64  
 1   carat    13485 non-null  float64
 2   cut      13485 non-null  object 
 3   color    13485 non-null  object 
 4   clarity  13485 non-null  object 
 5   depth    13485 non-null  float64
 6   table    13485 non-null  float64
 7   x        13485 non-null  float64
 8   y        13485 non-null  float64
 9   z        13485 non-null  float64
 10  city     13485 non-null  object 
dtypes: float64(6), int64(1), object(4)
memory usage: 1.1+ MB


In [17]:
media_x = diamonds_9['x'].mean()
media_y = diamonds_9['y'].mean()
media_z = diamonds_9['z'].mean()

In [18]:
diamonds_test['x'] = diamonds_test['x'].replace(0, media_x)
diamonds_test['y'] = diamonds_test['y'].replace(0, media_y)
diamonds_test['z'] = diamonds_test['z'].replace(0, media_z)

In [19]:
color_map_2 = {'J': 100,
             'I': 200,
             'H': 300,
             'G': 400,
             'F': 500,
             'E': 600,
             'D': 700}

In [20]:
clarity_map = {'I1': 1,
               'SI2': 2,
               'SI1': 3,
               'VS2': 4,
               'VS1': 5,
               'VVS2': 6,
               'VVS1': 7,
               'IF': 8}

In [21]:
cut_map_2 = {'Fair': 10,
           'Good': 50,
           'Very Good': 60,
           'Premium': 80,
           'Ideal': 100}

In [22]:
diamonds_test.drop('city', axis=1, inplace=True)

In [23]:
diamonds_test['cut'] = diamonds_test['cut'].map(cut_map_2)

In [24]:
diamonds_test['color'] = diamonds_test['color'].map(color_map_2)

In [25]:
diamonds_test['clarity'] = diamonds_test['clarity'].map(clarity_map)

In [26]:
diamonds_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13485 entries, 0 to 13484
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   id       13485 non-null  int64  
 1   carat    13485 non-null  float64
 2   cut      13485 non-null  int64  
 3   color    13485 non-null  int64  
 4   clarity  13485 non-null  int64  
 5   depth    13485 non-null  float64
 6   table    13485 non-null  float64
 7   x        13485 non-null  float64
 8   y        13485 non-null  float64
 9   z        13485 non-null  float64
dtypes: float64(6), int64(4)
memory usage: 1.0 MB


In [27]:
diamonds_test.drop('id', axis=1, inplace=True)

In [28]:
diamonds_9.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 40455 entries, 0 to 40454
Data columns (total 10 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   price    40455 non-null  int64  
 1   carat    40455 non-null  float64
 2   depth    40455 non-null  float64
 3   table    40455 non-null  float64
 4   x        40455 non-null  float64
 5   y        40455 non-null  float64
 6   z        40455 non-null  float64
 7   cut      40455 non-null  int64  
 8   color    40455 non-null  int64  
 9   clarity  40455 non-null  int64  
dtypes: float64(6), int64(4)
memory usage: 3.1 MB


In [29]:
diamonds_test.head()

carat  cut  color  clarity  depth  table     x     y     z
0   0.79   60    500        3   62.7   60.0  5.82  5.89  3.67
1   1.20  100    100        5   61.0   57.0  6.81  6.89  4.18
2   1.57   80    300        3   62.2   61.0  7.38  7.32  4.57
3   0.90   60    500        3   63.8   54.0  6.09  6.13  3.90
4   0.50   60    500        5   62.9   58.0  5.05  5.09  3.19

In [30]:
diamonds_test = diamonds_test[['carat', 'depth', 'table', 'x', 'y', 'z', 'cut', 'color', 'clarity']]

In [31]:
diamonds_test.head()

carat  depth  table     x     y     z  cut  color  clarity
0   0.79   62.7   60.0  5.82  5.89  3.67   60    500        3
1   1.20   61.0   57.0  6.81  6.89  4.18  100    100        5
2   1.57   62.2   61.0  7.38  7.32  4.57   80    300        3
3   0.90   63.8   54.0  6.09  6.13  3.90   60    500        3
4   0.50   62.9   58.0  5.05  5.09  3.19   60    500        5

In [105]:
model = RandomForestRegressor()

In [66]:
pred_test = model.predict(diamonds_test)

In [67]:
pred_test

array([2933.56, 5401.89, 9290.79, ..., 2944.62, 2231.04,  898.48])

In [68]:
pred_test = pd.DataFrame(pred_test, columns=['price'])

In [69]:
pred_test['id'] = range(0, len(pred_test))

In [70]:
pred_test = pred_test[['id', 'price']]

In [71]:
pred_test

id    price
0          0  2933.56
1          1  5401.89
2          2  9290.79
3          3  4133.37
4          4  1711.10
...      ...      ...
13480  13480  1848.12
13481  13481  2549.06
13482  13482  2944.62
13483  13483  2231.04
13484  13484   898.48

[13485 rows x 2 columns]

In [72]:
pred_test.to_csv('./submissions/sub_RandomForestRegressor_544.csv', index=False)

# GridSearchCV

In [73]:
from sklearn.model_selection import GridSearchCV

In [74]:
param_grid = {
    'n_estimators': [100, 300, 500],
    'max_depth': [None, 10, 20, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': ['auto', 'sqrt', 'log2'],
    'bootstrap': [True, False],
    'random_state': [42]
}

In [75]:
grid_search = GridSearchCV(model,
                           param_grid,
                           cv=5,
                           verbose=3,
                           scoring='neg_root_mean_squared_error',
                           n_jobs=-1)

In [76]:
grid_search.fit(X_9_train, y_9_train)     

print('\n')
print('Best hyperparameters: ', grid_search.best_params_, '\n')
print('Best score: ', -grid_search.best_score_, '\n')

Fitting 5 folds for each of 648 candidates, totalling 3240 fits


C:\Users\tete1\miniconda3\envs\jupyter_env\lib\site-packages\sklearn\model_selection\_validation.py:542: FitFailedWarning: 
1080 fits failed out of a total of 3240.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
165 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\tete1\miniconda3\envs\jupyter_env\lib\site-packages\sklearn\model_selection\_validation.py", line 890, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\tete1\miniconda3\envs\jupyter_env\lib\site-packages\sklearn\base.py", line 1344, in wrapper
    estimator._validate_params()
  File "C:\Users\tete1\miniconda3\envs\jupyter_env\lib\site-packages\sklearn\base.py", line 666, in _validate_params
    valida



Best hyperparameters:  {'bootstrap': False, 'max_depth': 20, 'max_features': 'sqrt', 'min_samples_leaf': 1, 'min_samples_split': 5, 'n_estimators': 500, 'random_state': 42} 

Best score:  567.7853224215603 



In [8]:
best_model = RandomForestRegressor(bootstrap= False,
                                    max_depth= 20,
                                    max_features= 'sqrt',
                                    min_samples_leaf= 1,
                                    min_samples_split= 5,
                                    n_estimators= 500,
                                    random_state= 42)

In [10]:
best_model.fit(X_9_train, y_9_train)

RandomForestRegressor(bootstrap=False, max_depth=20, max_features='sqrt',
                      min_samples_split=5, n_estimators=500, random_state=42)

In [12]:
best_pred_9 = best_model.predict(X_9_test)

In [14]:
best_rmse_9 = mean_squared_error(y_9_test, best_pred_9, squared=False)
best_rmse_9

C:\Users\tete1\miniconda3\envs\jupyter_env\lib\site-packages\sklearn\metrics\_regression.py:483: FutureWarning: 'squared' is deprecated in version 1.4 and will be removed in 1.6. To calculate the root mean squared error, use the function'root_mean_squared_error'.
  warnings.warn(


547.8917499406093

In [15]:
# Predice peor que sin los mejores parámetros, pero voy a generar submission

In [32]:
best_pred_test = best_model.predict(diamonds_test)

In [34]:
best_pred_test = pd.DataFrame(best_pred_test, columns=['price'])

In [35]:
best_pred_test['id'] = range(0, len(best_pred_test))

In [36]:
best_pred_test = best_pred_test[['id', 'price']]

In [37]:
best_pred_test.to_csv('./submissions/sub_RandomForestRegressor_547_GS.csv', index=False)

In [ ]:
# Score competi: 549